components/playlist/playlist_service.cc

```diff
@@ -242,6 +257,178 @@ void PlaylistService::ConfigureWebPrefsForBackgroundWebContents(
+  if (!playlists.contains(id)) {
+    LOG(ERROR) << __func__ << " playlist with id<" << id << "> not found";
+    std::move(callback).Run({});
+    return;
+  }
+  auto* playlist_dict = playlists.FindDict(id);
+  DCHECK(playlist_dict);
+
+  mojom::PlaylistPtr playlist = mojom::Playlist::New();
+  playlist->id = *playlist_dict->FindString(kPlaylistIDKey);
+  playlist->name = *playlist_dict->FindString(kPlaylistNameKey);
+  for (const auto& item_id_value :
+       *playlist_dict->FindList(kPlaylistItemsKey)) {
+    GetPlaylistItem(item_id_value.GetString(),
+                    base::BindOnce(
+                        [](std::vector<mojom::PlaylistItemPtr>* items,
+                           mojom::PlaylistItemPtr item) {
+                          items->push_back(std::move(item));
+                        },
+                        &playlist->items));
+  }
+
+  std::move(callback).Run(std::move(playlist));
+}
+
+void PlaylistService::GetAllPlaylistItems(
+    GetAllPlaylistItemsCallback callback) {
+  std::vector<mojom::PlaylistItemPtr> items;
+  for (const auto it : prefs_->GetDict(kPlaylistItemsPref)) {
+    items.push_back(GetPlaylistItemFromValue(it.second.GetDict()));
+  }
+
+  std::move(callback).Run(std::move(items));
+}
+
+void PlaylistService::GetPlaylistItem(const std::string& id,
+                                      GetPlaylistItemCallback callback) {
+  DCHECK(!id.empty());
+  const auto* item_value = prefs_->GetDict(kPlaylistItemsPref).FindDict(id);
+  DCHECK(item_value);
+  if (!item_value) {
+    std::move(callback).Run({});
+    return;
+  }
+
+  return std::move(callback).Run(GetPlaylistItemFromValue(*item_value));
+}
+
+void PlaylistService::AddMediaFilesFromPageToPlaylist(
+    const std::string& playlist_id,
+    const GURL& url) {
+  VLOG(2) << __func__ << " " << playlist_id << " " << url;
+  PlaylistDownloadRequestManager::Request request;
+  request.url_or_contents = url.spec();
+  request.callback = base::BindOnce(
+      &PlaylistService::AddMediaFilesFromItems, base::Unretained(this),
+      playlist_id.empty() ? GetDefaultSaveTargetListID() : playlist_id,
+      prefs_->GetBoolean(playlist::kPlaylistCacheByDefault));
+  download_request_manager_->GetMediaFilesFromPage(std::move(request));
+}
+
+void PlaylistService::AddMediaFilesFromOpenTabsToPlaylist(
+    const std::string& playlist_id) {
+  if (!delegate_) {
+    LOG(ERROR) << "Delegate should be set to use this method.";
+    return;
+  }
+
+  const bool should_cache = prefs_->GetBoolean(kPlaylistCacheByDefault);
+  for (auto* contents : delegate_->GetOpenTabsWebContentses())
+    AddMediaFilesFromContentsToPlaylist(playlist_id, contents, should_cache);
+}
+
+void PlaylistService::AddMediaFilesFromActiveTabToPlaylist(
+    const std::string& playlist_id) {
+  if (!delegate_) {
+    LOG(ERROR) << "Delegate should be set to use this method.";
+    return;
+  }
+
+  AddMediaFilesFromContentsToPlaylist(
+      playlist_id, delegate_->GetActiveTabWebContents(),
```

Comment:
using `FindLastActiveWithProfile` here is not safe because all messages from the renderer are async and the active tab can change.

Reply:
Done.